File: src/models/vision/type_hint.py
"""Type hint enumeration for visual element classification.

This module defines the TypeHint enum for coarse-grained visual type
classification by multimodal models.
"""

from enum import Enum


class TypeHint(str, Enum):
    """Coarse-grained visual type hint for element classification.

    These types represent visual features observable in screenshots,
    without behavioral inference. They are output by the multimodal
    model and later mapped to precise MenuItemType by the text model.

    Values:
        CLICKABLE_TEXT: Clickable text region (e.g., menu items)
        SWITCH: Switch/toggle control
        SLIDER: Slider control
        BUTTON: Button control
        ICON: Icon element (no text)
        INPUT_FIELD: Text input field
        TEXT: Plain text (non-interactive)
        IMAGE: Image element
    """

    CLICKABLE_TEXT = "clickable_text"
    SWITCH = "switch"
    SLIDER = "slider"
    BUTTON = "button"
    ICON = "icon"
    INPUT_FIELD = "input_field"
    TEXT = "text"
    IMAGE = "image"

    @classmethod
    def from_string(cls, value: str) -> 'TypeHint':
        """Create TypeHint from string with fuzzy matching.

        Args:
            value: String value to convert

        Returns:
            TypeHint enum instance

        Examples:
            >>> TypeHint.from_string("clickable_text")
            <TypeHint.CLICKABLE_TEXT: 'clickable_text'>
            >>> TypeHint.from_string("toggle")  # Fuzzy match
            <TypeHint.SWITCH: 'switch'>
        """
        value_lower = value.lower().strip()

        # Try exact match first
        try:
            return cls(value_lower)
        except ValueError:
            pass

        # Fuzzy mapping for common alternatives
        mapping = {
            'text': cls.TEXT,
            'clickable': cls.CLICKABLE_TEXT,
            'click': cls.CLICKABLE_TEXT,
            'toggle': cls.SWITCH,
            'checkbox': cls.SWITCH,
            'check': cls.SWITCH,
            'btn': cls.BUTTON,
            'input': cls.INPUT_FIELD,
            'field': cls.INPUT_FIELD,
            'img': cls.IMAGE,
            'picture': cls.IMAGE,
        }

        result = mapping.get(value_lower, cls.TEXT)
        return result

    @classmethod
    def values(cls) -> list[str]:
        """Get all enum values as a list of strings.

        Returns:
            List of valid TypeHint values
        """
        return [e.value for e in cls]

    @classmethod
    def is_valid(cls, value: str) -> bool:
        """Check if a string value is a valid TypeHint.

        Args:
            value: String value to validate

        Returns:
            True if value is a valid TypeHint
        """
        return value.lower() in cls.values()

    def is_interactive(self) -> bool:
        """Check if this type represents an interactive element.

        Returns:
            True if the element type is typically interactive
        """
        interactive_types = {
            self.CLICKABLE_TEXT,
            self.SWITCH,
            self.SLIDER,
            self.BUTTON,
            self.INPUT_FIELD,
        }
        return self in interactive_types

    def is_visual_only(self) -> bool:
        """Check if this type represents a visual-only element.

        Returns:
            True if the element type is visual-only (no interaction)
        """
        return self in {self.ICON, self.IMAGE, self.TEXT}

File: src/UniClaw.Core/Domain/Models/Vision/TypeHint.cs
using System.Text.Json.Serialization;

namespace UniClaw.Core.Domain.Models.Vision;

/// <summary>
/// 粗粒度的视觉元素分类枚举。无 Unknown（PRD §5.1）：未识别输入回落为 Text。
/// [JsonPropertyName] 锁定 snake_case 序列化名（P3：对齐 Python，与其他 3 个 Domain enum 一致）。
/// </summary>
public enum TypeHint
{
    /// <summary>可点击文本区域（如菜单项）</summary>
    [JsonPropertyName("clickable_text")] ClickableText,

    /// <summary>开关/切换控件</summary>
    [JsonPropertyName("switch")] Switch,

    /// <summary>滑块控件</summary>
    [JsonPropertyName("slider")] Slider,

    /// <summary>按钮控件</summary>
    [JsonPropertyName("button")] Button,

    /// <summary>图标元素（无文本）</summary>
    [JsonPropertyName("icon")] Icon,

    /// <summary>文本输入框</summary>
    [JsonPropertyName("input_field")] InputField,

    /// <summary>纯文本（非交互）</summary>
    [JsonPropertyName("text")] Text,

    /// <summary>图片元素</summary>
    [JsonPropertyName("image")] Image
}

/// <summary>
/// TypeHint 扩展方法
/// </summary>
public static class TypeHintExtensions
{
    /// <summary>所有合法枚举值</summary>
    public static IReadOnlyList<TypeHint> Values { get; } =
        (IReadOnlyList<TypeHint>)Enum.GetValues<TypeHint>();

    // ── 别名字典：8 精确枚举值 + 10 Python 别名 + 3 C# 扩展别名 ──
    private static readonly Dictionary<string, TypeHint> AliasMap = new()
    {
        // 8 精确枚举值
        ["clickable_text"] = TypeHint.ClickableText,
        ["switch"] = TypeHint.Switch,
        ["slider"] = TypeHint.Slider,
        ["button"] = TypeHint.Button,
        ["icon"] = TypeHint.Icon,
        ["input_field"] = TypeHint.InputField,
        ["text"] = TypeHint.Text,
        ["image"] = TypeHint.Image,

        // Python 别名
        ["clickable"] = TypeHint.ClickableText,   // Python alias
        ["click"] = TypeHint.ClickableText,       // Python alias (previously missing)
        ["toggle"] = TypeHint.Switch,             // Python alias
        ["checkbox"] = TypeHint.Switch,           // Python alias
        ["check"] = TypeHint.Switch,              // Python alias (previously missing)
        ["btn"] = TypeHint.Button,                // Python alias
        ["input"] = TypeHint.InputField,          // Python alias
        ["field"] = TypeHint.InputField,          // Python alias
        ["img"] = TypeHint.Image,                 // Python alias
        ["picture"] = TypeHint.Image,             // Python alias

        // C# 扩展别名（Python mapping dict 不含，但 C# 遍历场景实际遇到）
        ["seekbar"] = TypeHint.Slider,            // Android SeekBar
        ["edit"] = TypeHint.InputField,           // EditText
        ["textbox"] = TypeHint.InputField,        // AI variant
    };

    /// <summary>判断类型是否可交互</summary>
    public static bool IsInteractive(this TypeHint type) => type switch
    {
        TypeHint.ClickableText or TypeHint.Switch or TypeHint.Slider or
        TypeHint.Button or TypeHint.InputField => true,
        _ => false
    };

    /// <summary>判断类型是否仅为视觉元素（不可交互）</summary>
    public static bool IsVisualOnly(this TypeHint type) => !IsInteractive(type);

    /// <summary>判断枚举值是否在合法范围内</summary>
    public static bool IsValid(TypeHint value) => Enum.IsDefined(value);

    /// <summary>
    /// 从字符串创建 TypeHint：精确别名字典匹配 → 未识别回落 Text。
    /// 不返回 Unknown（已删除）。大小写不敏感，空格容错。
    /// </summary>
    public static TypeHint FromString(string value)
    {
        var key = value.ToLowerInvariant().Trim();
        return AliasMap.TryGetValue(key, out var result) ? result : TypeHint.Text;
    }

    /// <summary>
    /// 判断字符串能否被 FromString 成功解析（含别名），未识别值返回 false。
    /// </summary>
    public static bool IsValid(string value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        return AliasMap.ContainsKey(value.ToLowerInvariant().Trim());
    }

    /// <summary>
    /// 8 个 TypeHint 枚举值的精确规范名（snake_case，与 [JsonPropertyName] 一致）。
    /// </summary>
    private static readonly HashSet<string> CanonicalNames = new()
    {
        "clickable_text", "switch", "slider", "button",
        "icon", "input_field", "text", "image"
    };

    /// <summary>
    /// 判断字符串是否为 TypeHint 的精确规范名（8 个枚举值之一），区别于别名。
    /// IsValid(string) 含别名，IsCanonical(string) 仅精确值。
    /// </summary>
    public static bool IsCanonical(string value)
    {
        if (string.IsNullOrEmpty(value)) return false;
        return CanonicalNames.Contains(value.ToLowerInvariant().Trim());
    }
}
